src/utils/display-observer.js

```diff
@@ -0,0 +1,248 @@
+
+    /** @private @const {!Map<!Element, !Array<boolean>>} */
+    this.targetDisplayMap_ = new Map();
+  }
+
+  /** @override */
+  dispose() {
+    this.docObserver_.disconnect();
+    this.viewportObserver_.disconnect();
+    this.visibilityUnlisten_();
+    this.visibilityUnlisten_ = null;
+    this.targetObserverMultimap_.clear();
+    this.targetDisplayMap_.clear();
+  }
+
+  /**
+   * @param {!Element} target
+   * @param {function(boolean)} callback
+   */
+  observe(target, callback) {
+    let callbacks = this.targetObserverMultimap_.get(target);
+    if (!callbacks) {
+      callbacks = [];
+      this.targetObserverMultimap_.set(target, callbacks);
+      this.docObserver_.observe(target);
+      this.viewportObserver_.observe(target);
+    }
+    if (pushIfNotExist(callbacks, callback)) {
+      if (this.targetDisplayMap_.has(target)) {
+        setTimeout(() => {
+          const display = getDisplay(
+            this.targetDisplayMap_.get(target),
+            this.isDocVisible_
+          );
+          if (display != null) {
+            callCallbackNoInline(callback, display);
+          }
+        });
+      }
+    }
+  }
+
+  /**
+   * @param {!Element} target
+   * @param {function()} callback
+   */
+  unobserve(target, callback) {
+    const callbacks = this.targetObserverMultimap_.get(target);
+    if (!callbacks) {
+      return;
+    }
+    removeItem(callbacks, callback);
+    if (callbacks.length == 0) {
+      this.targetObserverMultimap_.delete(target);
+      this.targetDisplayMap_.delete(target);
+      this.docObserver_.unobserve(target);
+      this.viewportObserver_.unobserve(target);
+    }
+  }
+
+  /** @private */
+  docVisibilityChanged_() {
+    this.targetObserverMultimap_.forEach((callbacks, target) => {
+      const displays = this.targetDisplayMap_.get(target);
+      const oldDisplay = getDisplay(displays, !this.isDocVisible_);
+      const newDisplay = getDisplay(displays, this.isDocVisible_);
+      if (newDisplay != null && newDisplay !== oldDisplay) {
+        for (let k = 0; k < callbacks.length; k++) {
+          callCallbackNoInline(callbacks[k], newDisplay);
+        }
+      }
+    });
+  }
+
+  /**
+   * @param {!Array<!IntersectionObserverEntry>} entries
+   * @param {number} observer
+   * @private
+   */
+  observed_(entries, observer) {
+    const seen = new Set();
+    for (let i = entries.length - 1; i >= 0; i--) {
+      const {target, isIntersecting} = entries[i];
+      if (seen.has(target)) {
+        continue;
+      }
+      seen.add(target);
+      const callbacks = this.targetObserverMultimap_.get(target);
+      if (!callbacks) {
+        continue;
+      }
+      let displays = this.targetDisplayMap_.get(target);
+      if (!displays) {
+        displays = [];
+        this.targetDisplayMap_.set(target, displays);
+      }
+      const oldDisplay = getDisplay(displays, this.isDocVisible_);
+      displays[observer] = isIntersecting;
```

Comment:
Making `displays` this weird tuple thing makes the API 10x harder to understand. I had reread the file several times to finally see how it was being used. `displays` needs to be a struct with descriptive key names, eg:

```js
displays = {
  displayedByViewport: false,
  displayedByDoc: false,
};

Reply:
Hmm. I can do some renaming, but here's how I envisioned it:

1. There's an array of intersection observers. Small number of them K.
2. Thus there's an array of intersection observations indexed by each observer (same K).
3. To compute `isDisplayed` we need to do `OR` over this array: ANY observation is `true` => the result is `true`, ALL observations are `false` => the result is `false`, otherwise "not ready".
4. Since the K is fixed and small (2 in this case) I didn't want to pollute the code with `for` cycles.